spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/support/DefaultPublisherFactory.java

```diff
@@ -56,9 +59,17 @@
 
 	private final CredentialsProvider credentialsProvider;
 
-	public DefaultPublisherFactory(String projectId, ExecutorProvider executorProvider,
-			ChannelProvider channelProvider, CredentialsProvider credentialsProvider) {
-		this.projectId = projectId;
+	public DefaultPublisherFactory(GcpProjectIdProvider projectIdProvider,
+			ExecutorProvider executorProvider,
+			ChannelProvider channelProvider,
+			CredentialsProvider credentialsProvider) {
+		Assert.notNull(projectIdProvider, "The project ID provider can't be null.");
+		Assert.notNull(executorProvider, "The executor provider can't be null.");
+		Assert.notNull(channelProvider, "The channel provider can't be null.");
+		Assert.notNull(credentialsProvider, "The credentials provider can't be null.");
+
+		this.projectId = projectIdProvider.getProjectId()
+				.orElseThrow(MissingProjectIdException::new);
```

Comment:
Sorry that I wasn't clear at a first place regarding consistency between what you have introduced with `ProjectIdProvider` and all other existing providers in GCP.

I don't find the `Optional` as so cool feature. Just IMHO.

You missed my point about **versions** compatibility not between Spring and Spring Boot.
I mean that SF 5.0 is a but far away from SF 4.3 and if we have tests based on SF 5.0 that doesn't provide guarantee that we are compatible with SF 4.3 which is our base version in the `main`.
The same applies for the Spring Boot 1.5 and 2.0.

But I don't about incompatibility between Spring 4.3 and Boot 1.5. I'm just pretty sure that Boot 2.0 doesn't work in some places with SF 4.3 and opposite - Boot 1.5 doesn't work with SF 5.0 somewhere.

I don't mind to have Boot 2.0 tests to be sure that we don't need a new version of the project for compatibility, but the base line has to confirm that the `main` selected version works well by the provided out-of-the-box tests which is based on the same Frameworks versions.

That's how I see things.